File: backend/app/schemas/leave.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel


class LeaveAllowanceCreate(BaseModel):
    user_id: uuid.UUID
    year: int
    entitlement_days: int = 25
    carried_forward: int = 0


class LeaveAllowanceResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    year: int
    entitlement_days: int
    carried_forward: int
    used_days: int
    booked_days: int
    remaining: int = 0
    created_at: datetime
    model_config = {"from_attributes": True}


class LeaveRequestCreate(BaseModel):
    leave_type: str
    start_date: date
    end_date: date
    days: int
    notes: str | None = None


class LeaveRequestUpdate(BaseModel):
    leave_type: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    days: int | None = None
    status: str | None = None
    notes: str | None = None


class LeaveRequestResponse(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    user_id: uuid.UUID
    leave_type: str
    start_date: date
    end_date: date
    days: int
    status: str
    approved_by: uuid.UUID | None = None
    notes: str | None = None
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}

File: frontend/src/api/leave.ts
import { api } from './client';

export interface LeaveAllowance {
  id: string;
  workspace_id: string;
  user_id: string;
  year: number;
  entitlement_days: number;
  carried_forward: number;
  used_days: number;
  booked_days: number;
  remaining: number;
  created_at: string;
}

export interface LeaveRequest {
  id: string;
  workspace_id: string;
  user_id: string;
  leave_type: string;
  start_date: string;
  end_date: string;
  days: number;
  status: string;
  approved_by: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export const leaveApi = {
  // Allowances
  listAllowances: (workspaceId: string, params?: { user_id?: string; year?: number }) =>
    api.get<LeaveAllowance[]>(`/workspaces/${workspaceId}/leave/allowances`, { params }),

  createAllowance: (workspaceId: string, data: Partial<LeaveAllowance>) =>
    api.post<LeaveAllowance>(`/workspaces/${workspaceId}/leave/allowances`, data),

  // Requests
  listRequests: (workspaceId: string, params?: { user_id?: string; status?: string }) =>
    api.get<LeaveRequest[]>(`/workspaces/${workspaceId}/leave/requests`, { params }),

  createRequest: (
    workspaceId: string,
    data: {
      leave_type: string;
      start_date: string;
      end_date: string;
      days: number;
      notes?: string;
    },
  ) => api.post<LeaveRequest>(`/workspaces/${workspaceId}/leave/requests`, data),

  updateRequest: (workspaceId: string, requestId: string, data: Partial<LeaveRequest>) =>
    api.put<LeaveRequest>(`/workspaces/${workspaceId}/leave/requests/${requestId}`, data),
};
